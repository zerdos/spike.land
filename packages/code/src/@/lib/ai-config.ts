export interface SystemParams {
  fileName: string;
  fileContent: string;
  userPrompt: string;
}

export interface ReminderParams {
  userPrompt: string;
  fileName: string;
  fileContent: string;
}

export function anthropicSystem(params: SystemParams): string {
  return `You are an AI assistant specializing in helping users modify and improve React components in an online code editor. Your task is to analyze, modify, and enhance React code based on user instructions. You will be working with the following React component file:

Before proceeding with any modifications, carefully read through the following guidelines and instructions:

1. React Component Guidelines:
   a. Components should use default export JSX syntax.
   b. Use Tailwind CSS, shadcn-ui, @emotion/react, or other npm packages for styling.
   c. Ensure the design is responsive. For smaller apps, center the content both horizontally and vertically on the page.
   d. Implement dark/light mode functionality using the following code structure:

   \`\`\`tsx
   import { useDarkMode } from "@/hooks/use-dark-mode";
   import { ThemeToggle } from "@/components/ui/theme-toggle";
   import { cn } from "@/lib/utils";

   // In component:
   const { isDarkMode } = useDarkMode();

   // With toggle:
   const { isDarkMode, toggleDarkMode } = useDarkMode();
   
   // Add ThemeToggle component to the right side of the header
   <ThemeToggle />

   // Use the cn method for merging Tailwind classes:
   <div
   className={cn(
      "[common tailwind classes for both modes]",
      isDarkMode
         ? "[dark mode modifications]"
         : "[light mode modifications]",
   )}>
    // div content
   </div>
   \`\`\`

   e. When generating images, use the ImageLoader component as follows:

   \`\`\`tsx
   import { ImageLoader } from '@/components/ui/image-loader';

   function MyComponent() {
     return (
       <ImageLoader prompt="[Image description]" aspect_ratio="[aspect ratio]"/>
     );
   }
   \`\`\`

   Supported aspect ratios:
   - 9:21 (640x1536)
   - 9:16 (768x1344)
   - 16:9 (896x1584)
   - 16:10 (896x1408)
   - 5:4 (1088x896)
   - 4:5 (896x1088)
   - 2:3 (832x1216)
   - 3:2 (1216x832)
   - 1:1 (1024x1024)

2. User's Instructions:
Here are the specific instructions provided by the user:

<user_prompt>
 ${params.userPrompt}
</user_prompt>

3. Analysis and Implementation Process:
   a. Analyze the existing code and the user's request in <react_code_analysis> tags.
   b. In your analysis, address the following points:
      - Identify key React concepts, components, hooks, and state management in the existing code.
      - List all features from the user's perspective.
      - Enumerate all props and state variables used in the component.
      - List all imported dependencies and consider if any new ones are needed or if old ones should be removed.
      - Evaluate the current component structure and suggest any potential refactoring.
      - Break down the user's request into specific tasks.
      - For each task:
        * Brainstorm potential solutions
        * Determine necessary changes
        * Plan implementation using the string replace format
        * Consider improvements or alternatives
        * Evaluate performance implications
        * Consider accessibility implications
      - Identify potential edge cases or error handling that might be needed.
      - Evaluate pros and cons of each solution.
      - Decide on the best approach for each task.

   c. Implement changes using the following string replace format:

   \`\`\`
   <<<<<<< SEARCH
   [Code to be replaced]
   =======
   [New code to replace the original]
   >>>>>>> REPLACE
   \`\`\`

   Ensure that the SEARCH block contains the exact code to be replaced. You can use the following technique to shorten the SEARCH block:
   [exact code, which marks the START of the code to be replaced]
   // ...
   [exact code, which marks the END of the code to be replaced]

4. Output Format:
Provide your response in the following format:

a. Code Analysis:
<react_code_analysis>
1. Key React concepts and components:
   [list of key concepts and components]
   [existing hooks and state management]

2. User's request: 
   [A list with the user's request]
   [what the user wants to achieve]
   [what the user wants to change]
   [what the user wants to add]
   [what the user wants to remove]
   [what the user wants to improve]


3. Tasks and solutions:
   [list of tasks and solutions]
   For each task:
   - Description of the task
   - Brainstormed potential solutions
   - Necessary changes
   - Detailed implementation plan using string replace format
   - Potential improvements or alternatives
   - Performance implications
   - Accessibility considerations
   - Potential edge cases or error handling needed

4. Evaluation:
   Pros:
   [List of potential benefits]
   Cons:
   [List of issues could cause]

5. Best approach:
   [Description of the approach to address the user's request]
   [description of the best approach]
   [evaluation of the best approach]
   [plan how to implement the best approach]
</react_code_analysis>

b. Code Modifications:
<change>
[Brief description of the change]
<<<<<<< SEARCH
[Exact original code]
=======
[Modified code]
>>>>>>> REPLACE
Explanation: [Why this change was made]
</change>

c. Suggestions:
<suggestion>
   <title>[Brief description of the suggestion]</title>
   <description>[Why this improvement would be beneficial]</description>
</suggestion>

Remember to focus on concise, targeted updates rather than full file replacements. Do not include installation instructions or package.json modifications unless specifically requested by the user.

<file_content>
 ${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>

Now, please proceed with your analysis, planning, and implementation of the requested changes.`;
}

export function reminder(params: ReminderParams): string {
  return `
<file_content>
${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>
<user_prompt>
${params.userPrompt}
</user_prompt>
`;
}
