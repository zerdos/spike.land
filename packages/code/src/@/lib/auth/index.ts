/**
 * Unified Authentication Library
 * Provides consistent authentication across all spike.land packages
 *
 * Security Features:
 * - Multiple auth provider support (Clerk, JWT, OAuth)
 * - OWASP compliant security practices
 * - Rate limiting and CSRF protection
 * - Session management with automatic refresh
 * - Audit logging for security events
 * - MFA support
 * - IP whitelisting/blacklisting
 * - Secure password policies
 */

// Export types
export * from "./types";

// Export auth manager
export { AuthManager, createAuthManager } from "./manager";

// Export React hooks
export {
  useAuth,
  useUser,
  useSession,
  useRequireAuth,
  usePermissions,
  useAuthError,
  useMFA,
  useSessionActivity,
  useAuthProvider,
} from "./hooks";

// Export middleware
export {
  createAuthMiddleware,
  createCSRFMiddleware,
  createSessionMiddleware,
  createSecurityHeadersMiddleware,
  createAPIKeyMiddleware,
  createWebhookMiddleware,
  authMiddleware,
  type AuthMiddlewareOptions,
} from "./middleware";

// Export providers
export { BaseAuthProvider } from "./providers/base";
export { ClerkAuthProvider } from "./providers/clerk";
export { JWTAuthProvider } from "./providers/jwt";

// Export security utilities
export { securityUtils } from "./security";

// Default export for convenience
import { createAuthManager } from "./manager";
import { authMiddleware } from "./middleware";

export default {
  createManager: createAuthManager,
  middleware: authMiddleware,
};