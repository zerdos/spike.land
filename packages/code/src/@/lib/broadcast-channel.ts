const broadcastChannels = new Map<string, BroadcastChannel>();

export const getBroadcastChannel = (codeSpace: string) => {
  if (broadcastChannels.has(codeSpace)) {
    return broadcastChannels.get(codeSpace)!;
  }

  const bc = new BroadcastChannel(`codeSpace-${codeSpace}`);

  broadcastChannels.set(codeSpace, bc);

  return bc;
};
