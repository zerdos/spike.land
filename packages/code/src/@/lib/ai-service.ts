import { anthropicSystem, reminder } from "@/lib/ai-config";
import type { Message } from "@/lib/interfaces";
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

export interface AIServiceConfig {
  anthropicEndpoint: string;
  openAIEndpoint: string;
  gpt4oEndpoint: string;
  updateThrottleMs: number;
  retryWithClaudeEnabled: boolean;
  setIsStreaming: (isStreaming: boolean) => void;
}

export class AIService {
  private _config: AIServiceConfig;

  constructor(config: AIServiceConfig) {
    this._config = config;
  }

  public async sendToAnthropic(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new Anthropic({
        dangerouslyAllowBrowser: true,
        apiKey: 'DUMMY_API_KEY',
        baseURL: this._config.anthropicEndpoint,
      });

      const stream = await client.messages.create({
        max_tokens: 8192,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
        model: "claude-3-5-sonnet-latest",
        stream: true,
      });

      let content = "";
      for await (const messageStreamEvent of stream) {
        if (
          "type" in messageStreamEvent &&
          messageStreamEvent.type === "content_block_delta" &&
          messageStreamEvent.delta.type === "text_delta"
        ) {
          const chunkContent = messageStreamEvent.delta.text;
          if (chunkContent) {
            content += chunkContent;
            onUpdate(chunkContent);
          }
       
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public async sendToGpt4o(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new OpenAI({
        baseURL: this._config.gpt4oEndpoint,
        apiKey: 'DUMMY_API_KEY',
      });

      const stream = await client.chat.completions.create({
        model: "gpt-4o-mini",
        stream: true,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
      });

      let content = "";
      for await (const chunk of stream) {
        const chunkContent = chunk.choices[0]?.delta?.content;
        if (chunkContent) {
          content += chunkContent;
          onUpdate(chunkContent);
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public prepareClaudeContent(
    params: { content: string; messages: Message[]; codeNow: string; codeSpace: string; },
  ): string {
    const { content, messages, codeNow, codeSpace } = params;

    // Check for error messages that should be passed through directly
    if (content.toLowerCase().startsWith("i'm sorry")) {
      return content;
    }

    const lastMessage = messages[messages.length - 1];

    // If codeNow is different from the last message or there are no messages,
    // use anthropic system content
    // if (!lastMessage || lastMessage.content !== codeNow) {
      const result = anthropicSystem({
        fileName: codeSpace+".tsx",
        fileContent: codeNow,
        userPrompt: content,
      });
      return result;
    // }

    // If code hasn't changed, return reminder content
    // return reminder({
    //   userPrompt: content,
    //   fileName: codeSpace+ ".tsx",
    //   fileContent: codeNow,
    // });
  }
}
