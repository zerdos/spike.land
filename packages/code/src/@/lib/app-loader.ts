import { getCodeSpace } from "@/hooks/use-code-space";
import { getCodeSession } from "@/lib/code-session";
import type { ICode } from "@/lib/interfaces";
import { init } from "@/lib/tw-dev-setup";
import { SessionSynchronizer } from "@/services/SessionSynchronizer";

// Centralized type definitions
export interface AppContext {
  codeSpace: string;
  cSess: ICode;
  AppComponent: React.FC<AppComponentProps>;
}

export interface AppComponentProps {
  codeSpace: string;
  cSess: ICode;
}

/**
 * Initializes the application environment
 * Loads modules that should be available globally
 */
export const initializeAppEnvironment = async (): Promise<void> => {
  try {
    const [
      { enhancedFetch },
      { useArchive, useSpeedy },
    ] = await Promise.all([
      import("@/lib/enhanced-fetch"),
      import("@/lib/use-archive"),
    ]);

    // Make utilities available globally for debugging and development
    Object.assign(globalThis, {
      enhancedFetch,
      useArchive,
      useSpeedy,
    });

    console.log("App environment initialized successfully");
  } catch (error) {
    console.error("Error initializing app environment:", error);
    throw error;
  }
};

/**
 * Initializes a WebSocket connection for the specified code space
 */
export const initializeWebSocket = async (codeSpace: string): Promise<void> => {
  if (!codeSpace) {
    throw new Error("Cannot initialize WebSocket: codeSpace is null or undefined");
  }

  try {
    // Import using a relative path to match the project structure
    const { main } = await import("../../ws");
    await main(codeSpace);
    console.log(`WebSocket initialized for code space: ${codeSpace}`);
  } catch (error) {
    console.error("WebSocket initialization failed:", error);
    throw error;
  }
};

/**
 * Loads the application for a given code space
 * This handles initialization of all required components and sessions
 */
export const loadApp = async (pathname: string): Promise<AppContext | null> => {
  const codeSpace = getCodeSpace(pathname);

  if (!codeSpace) {
    console.error("Cannot load app: Invalid codeSpace from pathname", pathname);
    return null;
  }

  try {
    // Initialize tailwind
    await init();

    // Get code session
    const cSess = await getCodeSession(codeSpace);

    // Load the app component dynamically
    // Import using a relative path to match the project structure
    const { AppToRender } = await import("../../AppToRender");

    // Initialize app environment
    await initializeAppEnvironment();

    return {
      codeSpace,
      cSess,
      AppComponent: AppToRender,
    };
  } catch (error) {
    console.error("Error loading app:", error);
    return null;
  }
};

/**
 * Initializes a session synchronizer for the given code space and session
 */
export const initializeSessionSync = async (
  codeSpace: string,
  cSess: ICode,
): Promise<() => void> => {
  if (!codeSpace || !cSess) {
    console.error("Cannot initialize session sync: missing codeSpace or cSess");
    return () => {};
  }

  try {
    // Make cSess available globally for debugging
    Object.assign(globalThis, { cSess });

    // Initialize session synchronizer
    const sessionSync = new SessionSynchronizer(codeSpace);
    await sessionSync.init(await cSess.getSession());

    // Return the unsubscribe function
    return sessionSync.subscribe((sess) => {
      cSess.setSession(sess);
      // Note: setState callback would be handled by the component
    });
  } catch (error) {
    console.error("Error initializing session sync:", error);
    return () => {};
  }
};

/**
 * Utility to check if a route should initialize WebSocket
 */
export const shouldInitWebSocket = (pathname: string, codeSpace: string | null): boolean => {
  if (!pathname || !codeSpace) return false;
  return pathname === `/live/${codeSpace}/iframe`;
};
