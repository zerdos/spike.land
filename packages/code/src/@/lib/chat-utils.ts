import { replacePreservingWhitespace } from "@/lib/diff-utils";
import type { Message } from "@/lib/interfaces";

interface CodeModification {
  search: string;
  replace: string;
}

const CODE_MODIFICATION_REGEX = /<<<<<<< SEARCH[\s\S]*?=======[\s\S]*?>>>>>>> REPLACE/g;
export const SEARCH_REPLACE_MARKERS = {
  SEARCH_START: "<<<<<<< SEARCH",
  SEPARATOR: "=======",
  REPLACE_END: ">>>>>>> REPLACE",
} as const;

/**
 * Adds or updates a message in the messages array
 * @param messages - The current array of messages
 * @param newMessage - The message to add or update
 * @returns A new array of messages with the new message added or updated
 */
export function messagesPush(
  messages: Message[],
  newMessage: Message,
): Message[] {
  if (!newMessage.role) {
    throw new Error("Message must have a role");
  }

  const messagesCopy = JSON.parse(JSON.stringify(messages)) as Message[];
  const newId = newMessage.id || Date.now().toString();

  if (!messagesCopy.length) {
    return [{ ...newMessage, id: newId }];
  }

  const lastMessage = messagesCopy[messagesCopy.length - 1];

  if (lastMessage.role === newMessage.role && newMessage.role === "assistant") {
    messagesCopy[messagesCopy.length - 1] = {
      ...lastMessage,
      id: newMessage.id,
      content: (newMessage.content as string).startsWith(lastMessage.content as string)
        ? newMessage.content as string
        : lastMessage.content as string + newMessage.content as string,
    };
    return messagesCopy;
  }

  return [...messagesCopy, { ...newMessage, id: newId }];
}

/**
 * Formats code as a section with a title
 */
export const formatCodeAsSection = (
  codeSpace: string,
  code: string,
): string => `
# ${codeSpace}.tsx

\`\`\`tsx
${code}
\`\`\`
`;

/**
 * Formats a code modification into the standard format
 */
function formatCodeModification(mod: CodeModification): string {
  return `<<<<<<< SEARCH
${mod.search}
=======
${mod.replace}
>>>>>>> REPLACE`;
}

/**
 * Parses a code block to extract search and replace parts
 */
function parseCodeBlock(block: string): string | null {
  const cleanBlock = block.trim();
  const lines = cleanBlock.split("\n");

  // Remove code fence markers if present
  if (lines[0].includes("```")) {
    lines.shift();
  }
  if (lines[lines.length - 1].includes("```")) {
    lines.pop();
  }

  // Handle standard SEARCH/REPLACE format
  if (cleanBlock.includes(SEARCH_REPLACE_MARKERS.SEARCH_START)) {
    const content = lines.join("\n")
      .replace(/<<<<<<< SEARCH|>>>>>>> REPLACE/g, "")
      .split(SEARCH_REPLACE_MARKERS.SEPARATOR);

    if (content.length === 2) {
      return formatCodeModification({
        search: content[0].trim(),
        replace: content[1].trim(),
      });
    }
    return null;
  }

  // Handle alternative format with multiple separators
  const separatorIndices = [];
  let startIdx = 0;
  const fullContent = lines.join("\n");
  
  while (true) {
    const idx = fullContent.indexOf(SEARCH_REPLACE_MARKERS.SEPARATOR, startIdx);
    if (idx === -1) break;
    separatorIndices.push(idx);
    startIdx = idx + SEARCH_REPLACE_MARKERS.SEPARATOR.length;
  }

  if (separatorIndices.length > 0) {
    const search = fullContent.substring(0, separatorIndices[0]).trim();
    const replace = fullContent.substring(separatorIndices[0] + SEARCH_REPLACE_MARKERS.SEPARATOR.length, separatorIndices[1] || fullContent.length).trim();
    return formatCodeModification({
      search,
      replace,
    });
  }

  return null;
}

/**
 * Extracts code modifications from a response string
 */
export const extractCodeModification = (response: string): string[] => {
  console.debug("extractCodeModification - Response length:", response.length);
  console.debug(
    "extractCodeModification - Contains SEARCH marker:",
    response.includes(SEARCH_REPLACE_MARKERS.SEARCH_START),
  );
  console.debug(
    "extractCodeModification - Contains SEPARATOR marker:",
    response.includes(SEARCH_REPLACE_MARKERS.SEPARATOR),
  );
  console.debug(
    "extractCodeModification - Contains REPLACE marker:",
    response.includes(SEARCH_REPLACE_MARKERS.REPLACE_END),
  );

  const modifications: string[] = [];

  // Extract modifications from regular format
  const regexMatches = response.match(CODE_MODIFICATION_REGEX) || [];
  console.debug("extractCodeModification - Regex matches count:", regexMatches.length);

  if (regexMatches.length > 0 && regexMatches[0]) {
    console.debug(
      "extractCodeModification - First regex match:",
      regexMatches[0].substring(0, 50) + (regexMatches[0].length > 50 ? "..." : ""),
    );
  } else {
    console.debug("extractCodeModification - Regex pattern used:", CODE_MODIFICATION_REGEX);
    console.debug(
      "extractCodeModification - Response sample:",
      response.substring(0, 100) + (response.length > 100 ? "..." : ""),
    );
  }

  modifications.push(...regexMatches);

  // Extract modifications from code blocks
  const codeBlockMatches = response.match(/```[\s\S]*?```/g) || [];
  console.debug("extractCodeModification - Code block matches count:", codeBlockMatches.length);

  codeBlockMatches.forEach((block, index) => {
    console.debug(`extractCodeModification - Code block ${index + 1} length:`, block.length);
    const modification = parseCodeBlock(block);
    if (modification) {
      console.debug(`extractCodeModification - Successfully parsed code block ${index + 1}`);
      modifications.push(modification);
    } else {
      console.warn(`extractCodeModification - Failed to parse code block ${index + 1}`);
      console.debug(
        `extractCodeModification - Code block ${index + 1} content:`,
        block.substring(0, 50) + (block.length > 50 ? "..." : ""),
      );
    }
  });

  console.debug("extractCodeModification - Total modifications extracted:", modifications.length);
  return modifications;
};

/**
 * Parses a code modification string into search and replace parts
 */
function parseModification(mod: string): CodeModification | null {
  console.debug("parseModification - Modification length:", mod.length);
  console.debug(
    "parseModification - Contains SEARCH marker:",
    mod.includes(SEARCH_REPLACE_MARKERS.SEARCH_START),
  );
  console.debug(
    "parseModification - Contains SEPARATOR marker:",
    mod.includes(SEARCH_REPLACE_MARKERS.SEPARATOR),
  );
  console.debug(
    "parseModification - Contains REPLACE marker:",
    mod.includes(SEARCH_REPLACE_MARKERS.REPLACE_END),
  );

  // Log the first few characters to help identify format issues
  console.debug(
    "parseModification - Modification start:",
    JSON.stringify(mod.substring(0, 50) + (mod.length > 50 ? "..." : "")),
  );

  const parts = mod
    .replace(/<<<<<<< SEARCH|>>>>>>> REPLACE/g, "")
    .split(SEARCH_REPLACE_MARKERS.SEPARATOR);

  console.debug("parseModification - Split parts count:", parts.length);

  if (parts.length === 2) {
    const search = parts[0].trim();
    const replace = parts[1].trim();

    console.debug("parseModification - Search part length:", search.length);
    console.debug("parseModification - Replace part length:", replace.length);

    return {
      search,
      replace,
    };
  }

  console.warn("parseModification - Failed to parse modification, parts count:", parts.length);
  return null;
}

/**
 * Loads messages from localStorage for a given code space
 */
export const loadMessages = (codeSpace: string): Message[] => {
  if (!codeSpace) {
    throw new Error("Code space must be provided");
  }

  const key = `chatMessages-${codeSpace}`;
  let rawMessages: Message[] = [];

  try {
    rawMessages = JSON.parse(localStorage.getItem(key) || "[]");
  } catch (error) {
    console.error("Error loading messages:", error);
    return [];
  }

  const validMessages = rawMessages.filter((m) => !!m.role);

  return validMessages.reduce((acc, current, index) => {
    if (index === 0 || current.role !== validMessages[index - 1].role) {
      acc.push(current);
    }
    return acc;
  }, [] as Message[]);
};

/**
 * Applies code modifications to the current code
 */
function applyCodeModifications(
  code: string,
  modifications: string[],
  applyAll = true,
): string {
  console.debug("applyCodeModifications - Code length:", code.length);
  console.debug("applyCodeModifications - Modifications count:", modifications.length);
  console.debug("applyCodeModifications - Apply all:", applyAll);

  try {
    let result = code;
    const modsToApply = applyAll ? modifications : modifications.slice(0, 1);
    console.debug("applyCodeModifications - Mods to apply count:", modsToApply.length);

    modsToApply.forEach((mod, index) => {
      console.debug(`applyCodeModifications - Applying modification ${index + 1}`);
      const parsed = parseModification(mod);

      if (parsed) {
        console.debug(
          `applyCodeModifications - Search part ${index + 1} length:`,
          parsed.search.length,
        );
        console.debug(
          `applyCodeModifications - Replace part ${index + 1} length:`,
          parsed.replace.length,
        );

        // Check if search exists in result (exact match)
        const exactMatch = result.includes(parsed.search);
        console.debug(
          `applyCodeModifications - Exact match for modification ${index + 1}:`,
          exactMatch,
        );

        // Apply the modification
        const beforeLength = result.length;
        result = replacePreservingWhitespace(
          result,
          parsed.search,
          parsed.replace,
        );
        const afterLength = result.length;

        // Check if the modification was applied
        const changed = beforeLength !== afterLength || result !== code;
        console.debug(`applyCodeModifications - Modification ${index + 1} applied:`, changed);
        console.debug(`applyCodeModifications - Length change:`, afterLength - beforeLength);
      } else {
        console.warn(`applyCodeModifications - Failed to parse modification ${index + 1}`);
      }
    });

    // Check if any changes were made
    const changed = result !== code;
    console.debug("applyCodeModifications - Overall changes made:", changed);

    return result;
  } catch (error) {
    console.error("Error applying code modifications:", error);
    console.error("Error details:", error instanceof Error ? error.message : String(error));
    console.error("Error stack:", error instanceof Error ? error.stack : "No stack trace");
    return code;
  }
}

/**
 * Updates code by applying all search/replace modifications
 */
export const updateSearchReplace = (
  instructions: string,
  codeNow: string,
): string => {
  console.debug("updateSearchReplace - Instructions length:", instructions.length);
  console.debug("updateSearchReplace - Code length:", codeNow.length);

  const modifications = extractCodeModification(instructions);
  console.debug("updateSearchReplace - Extracted modifications count:", modifications.length);

  if (modifications.length === 0) {
    console.warn("updateSearchReplace - No modifications extracted from instructions");
    console.debug("updateSearchReplace - Instructions:", instructions);
    return codeNow;
  }

  // Log each modification
  modifications.forEach((mod, index) => {
    console.debug(
      `updateSearchReplace - Modification ${index + 1}:`,
      mod.substring(0, 100) + (mod.length > 100 ? "..." : ""),
    );

    // Parse the modification to get search and replace parts
    const parsed = parseModification(mod);
    if (parsed) {
      console.debug(`updateSearchReplace - Search part ${index + 1} length:`, parsed.search.length);
      console.debug(
        `updateSearchReplace - Replace part ${index + 1} length:`,
        parsed.replace.length,
      );

      // Check if search exists in code (exact match)
      const exactMatch = codeNow.includes(parsed.search);
      console.debug(`updateSearchReplace - Exact match for modification ${index + 1}:`, exactMatch);

      // Check if search exists in code (ignoring whitespace)
      const codeNoWS = codeNow.replace(/\s+/g, "");
      const searchNoWS = parsed.search.replace(/\s+/g, "");
      console.debug(
        `updateSearchReplace - Match ignoring whitespace for modification ${index + 1}:`,
        codeNoWS.includes(searchNoWS),
      );
    } else {
      console.warn(`updateSearchReplace - Failed to parse modification ${index + 1}`);
    }
  });

  const result = applyCodeModifications(codeNow, modifications, true);

  // Check if any changes were made
  const changed = result !== codeNow;
  console.debug("updateSearchReplace - Changes made:", changed);

  return result;
};

/**
 * Updates code by applying only the first search/replace modification
 */
export const replaceFirstCodeMod = (
  instructions: string,
  codeNow: string,
): string => {
  const modifications = extractCodeModification(instructions);
  return applyCodeModifications(codeNow, modifications, false);
};
