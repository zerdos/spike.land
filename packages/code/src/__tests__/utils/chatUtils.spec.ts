import { extractCodeModification, loadMessages, updateSearchReplace } from "@/lib/chat-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";

describe("updateSearchReplace", () => {
  it('should return the original code if instructions do not contain "SEARCH"', () => {
    const instructions = "Some random instructions";
    const codeNow = "const a = 1;";
    const result = updateSearchReplace(instructions, codeNow);
    expect(result).toBe(codeNow);
  });

  it("should return modified code when instructions contain valid search and replace markers", () => {
    const instructions = `
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
    `;
    const codeNow = "const a = 1;";
    const result = updateSearchReplace(instructions, codeNow);
    expect(result).toBe("const a = 2;");
  });

  it("should handle multiple modifications", () => {
    const inst = `
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE

      <<<<<<< SEARCH
      let b = 3;
      =======
      let b = 4;
      >>>>>>> REPLACE
    `;
    const codeNow = "const a = 1;\nlet b = 3;";
    const instructions = extractCodeModification(inst);
    expect(instructions).toMatchSnapshot();

    const result = updateSearchReplace(inst, codeNow);
    expect(result).toBe("const a = 2;\nlet b = 4;");
  });

  it("should return the original code if an error occurs", () => {
    const instructions = `
      <<<<<<< SEARCH
      const asqqaa = 1;
      =======
      const a = 2;
      >s>>>>>> REPLACE
    `;
    const codeNow = "const a = 1; ";

    const result = updateSearchReplace(instructions, codeNow);
    expect(result).toBe(codeNow);
  });

  it("should handle broken search replace blocks", () => {
    const codeNow = `
    const a = 1;
    const b = 2;
    const c = 3;
    `;

    const instructions = `
\`\`\`tsx
    const a = 1;
    =======
    const a = 10;
    =======
    fooo bar blalalal
\`\`\`
    `;

    const result = updateSearchReplace(instructions, codeNow);
    expect(result).toBe(`
    const a = 10;
    const b = 2;
    const c = 3;
    `);
  });

  it("should handle multiple broken search replace blocks", () => {
    const codeNow = `
    const a = 1;
    const b = 2;
    const c = 3;
    `;

    const instructions = `
\`\`\`tsx
    const a = 1;
    =======
    const a = 10;
    =======
    fooo bar blalalal
\`\`\`


some text between
\`\`\`tsx
    const b = 2;
    =======
    const b = 20;
 \`\`\`
 
 

\`\`\`tsx
    <<<<<<< SEARCH
    const c = 3;
    =======
    const c = 30;
\`\`\`
\`\`\`tsx
    <<<<<<< SEARCH
    This is a broken block
    =======
    This should be ignored
\`\`\`
    `;

    const result = updateSearchReplace(instructions, codeNow);
    expect(result).toBe(`
    const a = 10;
    const b = 20;
    const c = 30;
    `);
  });
});

describe("extractCodeModification", () => {
  it("should extract code modifications from response", () => {
    const response = `
      Some text
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
      More text
    `;
    const result = extractCodeModification(response);
    expect(result).toHaveLength(1);
    expect(result[0]).toContain("<<<<<<< SEARCH");
    expect(result[0]).toContain("=======");
    expect(result[0]).toContain(">>>>>>> REPLACE");
  });

  it("should handle multiple code modifications", () => {
    const response = `
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
      Some text
      <<<<<<< SEARCH
      let b = 3;
      =======
      let b = 4;
      >>>>>>> REPLACE
    `;
    const result = extractCodeModification(response);
    expect(result).toHaveLength(2);
  });

  it("should handle code blocks with search and replace markers", () => {
    const response = `
      \`\`\`tsx
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
      \`\`\`
    `;
    const result = extractCodeModification(response);
    expect(result).toHaveLength(2);
    expect(result[0]).toContain("<<<<<<< SEARCH");
    expect(result[0]).toContain("=======");
    expect(result[0]).toContain(">>>>>>> REPLACE");
    expect(result[1]).toContain("const a = 1;");
    expect(result[1]).toContain("=======");
    expect(result[1]).toContain("const a = 2;");
  });
});

describe("loadMessages", () => {
  beforeEach(() => {
    vi.spyOn(Storage.prototype, "getItem");
    localStorage.clear();
  });

  it("should load messages from localStorage", () => {
    const codeSpace = "testSpace";
    const messages = [
      { role: "user", content: "Hello" },
      { role: "assistant", content: "Hi there" },
    ];
    localStorage.setItem(`chatMessages-${codeSpace}`, JSON.stringify(messages));

    const result = loadMessages(codeSpace);
    expect(result).toEqual(messages);
    expect(localStorage.getItem).toHaveBeenCalledWith(
      `chatMessages-${codeSpace}`,
    );
  });

  it("should filter out messages without a role", () => {
    const codeSpace = "testSpace";
    const messages = [
      { role: "user", content: "Hello" },
      { content: "This should be filtered out" },
      { role: "assistant", content: "Hi there" },
    ];
    localStorage.setItem(`chatMessages-${codeSpace}`, JSON.stringify(messages));

    const result = loadMessages(codeSpace);
    expect(result).toEqual([
      { role: "user", content: "Hello" },
      { role: "assistant", content: "Hi there" },
    ]);
  });

  it("should remove consecutive messages with the same role", () => {
    const codeSpace = "testSpace";
    const messages = [
      { role: "user", content: "Hello" },
      { role: "user", content: "How are you?" },
      { role: "assistant", content: "Hi there" },
      { role: "assistant", content: "I'm doing well" },
      { role: "user", content: "Great!" },
    ];
    localStorage.setItem(`chatMessages-${codeSpace}`, JSON.stringify(messages));

    const result = loadMessages(codeSpace);
    expect(result).toEqual([
      { role: "user", content: "Hello" },
      { role: "assistant", content: "Hi there" },
      { role: "user", content: "Great!" },
    ]);
  });

  it("should return an empty array if no messages are found", () => {
    const codeSpace = "emptySpace";
    const result = loadMessages(codeSpace);
    expect(result).toEqual([]);
  });
});
