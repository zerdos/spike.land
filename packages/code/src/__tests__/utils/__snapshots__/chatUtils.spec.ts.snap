// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`updateSearchReplace > should handle multiple modifications 1`] = `
[
  "<<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE",
  "<<<<<<< SEARCH
      let b = 3;
      =======
      let b = 4;
      >>>>>>> REPLACE",
]
`;
